File: Java/src/Maximal Rectangle.java
public class Solution {
    public int maximalRectangle(char[][] matrix) {
        int ret = 0;
        if (matrix==null||matrix.length==0||matrix[0].length==0)  return ret;
        int[] height = new int[matrix[0].length];
        for (int i = 0; i< matrix.length; i++){
            for (int j = 0; j < matrix[0].length; j++){
                if (matrix[i][j] == '1'){
                    height[j] = height[j]+1;
                }
                else{
                    height[j] = 0;
                }
            }
            ret = Math.max(ret, largestRectangleArea(height));
        }
        return ret;
    }
    
    public int largestRectangleArea(int[] height) {
        int length = height.length;
        Stack<Integer> stack = new Stack<>();
        int ret = 0;
        int hi;
        int s; 

        for (int i = 0; i <= length; i++){
           if (i == length){
                hi = 0;
            }
            else{
                hi = height[i];
            }
            while (!stack.isEmpty() && height[stack.peek()] >= hi) {
                int h = height[stack.pop()];
                if (!stack.isEmpty()){
                    s = stack.peek();
                }
                else{
                    s = -1;
                }
                ret = Math.max(ret, h*(i-s-1));
            }
            stack.push(i);
        }
        return ret;
    }
}

File: Python/Maximal Rectangle.py
class Solution:
    # @param matrix, a list of lists of 1 length string
    # @return an integer
    def maximalRectangle(self, matrix):
        ret = 0
        if not matrix:
            return 0
        height = len(matrix)
        width = len(matrix[0])
        h = [0]*width

        for i in range(height):
            for j in range(width):
                if matrix[i][j] == '1':
                    h[j] = h[j] + 1
                else:
                    h[j] = 0
            ret = max(self.largestRectangleArea(h), ret)
        return ret

    def largestRectangleArea(self, height):
        ret = 0
        height.append(0)
        stack = []
        
        for i in range(len(height)):
            while len(stack) > 0 and height[stack[-1]] >= height[i]:
                h = height[stack.pop()]
                if stack:
                    s = stack[-1]
                else:
                    s = -1
                if  h * (i - s - 1 ) > ret:
                    ret = h * (i - s - 1 )
            stack.append(i)
        return ret
        